### PR TITLE
[JENKINS-58798] Hotfix/jenkins rebuild

### PR DESCRIPTION
**Pull request by baymac**
* Jenkins rebuild is added as branch property. The impl is entirely inspired from https://github.com/jenkinsci/github-pr-comment-build-plugin

* Also found a bug in the way `getTrustedRevision` impl was done. Having a transient variable caused problem when externally triggering a build. So added a HashSet to store the trusted mr heads during indexing in `GitLabSCMSource` and use the hashset in `getTrustedRevision`.

### CODE REVIEW

**Comment by baymac** [comment]
The new impl of `getTrustedRevision` allows us to enable check for updated members list.

**Comment by baymac** [comment]
The problem with the trusted revision is fixed. Now the problem with the comment trigger is that when the Jenkins is restarted then comment trigger doesn't work. Wondering what could be the problem? :/

**Comment by baymac** [comment]
I will try change this impl from branch property to a SCM trait later. Branch Property has some issues https://github.com/jenkinsci/github-pr-comment-build-plugin/issues/1.

**Comment by baymac** [comment]
The new trait makes it compatible with Folder organization as well! But the isMatch triggers the build twice for some reason. sweat

**Comment by baymac** [comment]
The `jenkins rebuild` trigger works even when Jenkins is restarted.

**Comment by baymac** [comment]
The duplicate build of navigator projects is happening as the `SCMOwners.all()` returns navigator as one entry and the sources inside the navigator as different entries. I need to find a way to skip sources which are children of a navigator.

**Comment by baymac** [comment]
A new hack enables us to skip scmsources which are part of scmnavigators thereby not duplicating build triggers.

File: src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabMergeRequestCommentTrigger.java [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+    public void isMatch() {
+        if(getPayload().getObjectAttributes().getNoteableType().equals(NoteEvent.NoteableType.MERGE_REQUEST)) {
+            Integer mergeRequestId = getPayload().getMergeRequest().getIid();
+            final Pattern mergeRequestJobNamePattern = Pattern.compile("^MR-" + mergeRequestId + "\\b.*$",
+                    Pattern.CASE_INSENSITIVE);
+            final String commentBody = getPayload().getObjectAttributes().getNote();
+            final String commentUrl = getPayload().getObjectAttributes().getUrl();
+            ACL.impersonate(ACL.SYSTEM, () -> {
+                boolean jobFound = false;
+                for (final SCMSourceOwner owner : SCMSourceOwners.all()) {
```
Comment (casz) [inline src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabMergeRequestCommentTrigger.java]: I suggest you try a lambda that filter `SCMSourceOwners` and map them to `SCMsources` and then filter `SCMSources`

Comment (baymac) [reply src/main/java/io/jenkins/plugins/gitlabbranchsource/GitLabMergeRequestCommentTrigger.java]:
```java
@Override
    public void isMatch() {
        if(getPayload().getObjectAttributes().getNoteableType().equals(NoteEvent.NoteableType.MERGE_REQUEST)) {
            Integer mergeRequestId = getPayload().getMergeRequest().getIid();
            final Pattern mergeRequestJobNamePattern = Pattern.compile("^MR-" + mergeRequestId + "\\b.*$",
                    Pattern.CASE_INSENSITIVE);
            final String commentBody = getPayload().getObjectAttributes().getNote();
            final String commentUrl = getPayload().getObjectAttributes().getUrl();
            ACL.impersonate(ACL.SYSTEM, () -> {
                boolean jobFound = false;
                StreamSupport.stream(SCMSourceOwners.all().spliterator(), false)
                        .filter(owner -> !owner.getFullDisplayName().contains(" » "))
                        .map(SCMSourceOwner::getSCMSources)
                        .filter(scmSource -> scmSource instanceof GitLabSCMSource)
                        .forEach(scmSource -> {
                        GitLabSCMSource gitLabSCMSource = (GitLabSCMSource) scmSource;
                        final GitLabSCMSourceContext sourceContext = new GitLabSCMSourceContext(null, SCMHeadObserver.none())
                                .withTraits(gitLabSCMSource.getTraits());
                        if (!sourceContext
                                .mrCommentTriggerEnabled()) {
                            return;
                        }
                        if (gitLabSCMSource.getProjectId() == getPayload().getMergeRequest().getTargetProjectId()) {
                            for (Job<?, ?> job : owner.getAllJobs()) {
                                if (mergeRequestJobNamePattern.matcher(job.getName()).matches()) {
                                    String expectedCommentBody = sourceContext.getCommentBody();
                                        Pattern pattern = Pattern.compile(expectedCommentBody,
                                                Pattern.CASE_INSENSITIVE | Pattern.DOTALL);
                                        if (commentBody == null || pattern.matcher(commentBody).matches()) {
                                            ParameterizedJobMixIn.scheduleBuild2(job, 0,
                                                    new CauseAction(new GitLabMergeRequestCommentCause(commentUrl)));
                                            LOGGER.log(Level.INFO,
                                                    "Triggered build for {0} due to MR comment on {1}",
                                                    new Object[] {
                                                            job.getFullName(),
                                                            getPayload().getProject().getPathWithNamespace()
                                                    }
                                            );
                                        } else {
                                            LOGGER.log(Level.INFO,
                                                    "MR comment does not match the trigger build string ({0}) for {1}",
                                                    new Object[] { expectedCommentBody, job.getFullName() }
                                            );
                                        }
                                        break;
                                    }
                                    jobFound = true;
                                }
                            }
                        });
                if (!jobFound) {
                    LOGGER.log(Level.INFO, "MR comment on {0} did not match any job",
                            new Object[] {
                                    getPayload().getProject().getPathWithNamespace()
                            }
                    );
                }
            });
        }
    }
```

There are two issues with this impl:

1. `jobFound` variable should be final or effectively final
2. Cannot access `owner` inside `forEach` of SCMSources.

**Comment by baymac** [comment]
@casz I tried using streams with filters and maps but couldn't make it work. I am giving it up for now. I made couple of other changes to perform rebuild from only trusted users and modified GitLabSCMSource to always supply the latest members with their access level.

**Pull request merged**